util/http/retry.go

```diff
@@ -42,82 +43,62 @@ const (
 	DefaultMaxRetries = 8
 	// DefaultMinWaitMsec is the default minimum number of milliseconds between attempts. See `RetryConfig.MinWait`.
 	DefaultMinWaitMsec = 30
-	// DefaultMaxWaitMsec is the default maxmimum number of millisends between attempts. See `RetryConfig.MaxWait`.
+	// DefaultMaxWaitMsec is the default maximum number of milliseconds between attempts. See `RetryConfig.MaxWait`.
 	DefaultMaxWaitMsec = 300_000
 )
 
-// RetryConfig represents the settings for retries in a retryable http client.
-type RetryConfig struct {
-	// MaxRetries is the maximum number of retries before giving up on a retryable request.
-	// This does not include the initial request so the total number of attempts will be MaxRetries + 1.
-	MaxRetries int
-	// MinWait is the minimum wait time between attempts. The actual wait time is governed by the BackoffStrategy,
-	// but the wait time will never be shorter than this duration.
-	MinWait time.Duration
-	// MaxWait is the maximum wait time between attempts. The actual wait time is governed by the BackoffStrategy,
-	// but the wait time will never be longer than this duration.
-	MaxWait time.Duration
-}
+// ParseRetryableClientConfig parses RetryableHTTPClientConfig making sure all 0'd values
+// are set to their appropriate defaults.
+func ParseRetryableClientConfig(cfg config.RetryableHTTPClientConfig) config.RetryableHTTPClientConfig {
+	if cfg.DialTimeout == 0 {
+		cfg.DialTimeout = DefaultDialTimeoutMsec
+	}
+	if cfg.ResponseHeaderTimeout == 0 {
+		cfg.ResponseHeaderTimeout = DefaultResponseHeaderTimeoutMsec
 
-// TimeoutConfig represents the settings for timeout at various points in a request lifecycle in a retryable http client.
-type TimeoutConfig struct {
-	// DialTimeout is the maximum duration that connection can take before a request attempt is timed out.
-	DialTimeout time.Duration
-	// ResponseHeaderTimeout is the maximum duration waiting for response headers before a request attempt is timed out.
-	// This starts after the entire request body is uploaded to the remote endpoint and stops when the request headers
-	// are fully read. It does not include reading the body.
-	ResponseHeaderTimeout time.Duration
-	// RequestTimeout is the maximum duration before the entire request attempt is timed out. This starts when the
-	// client starts the connection attempt and ends when the entire response body is read.
-	RequestTimeout time.Duration
-}
+	}
+	if cfg.RequestTimeout == 0 {
+		cfg.RequestTimeout = DefaultRequestTimeoutMsec
 
-// RetryableClientConfig is the complete config for a retryable http client
-type RetryableClientConfig struct {
-	TimeoutConfig
-	RetryConfig
-}
+	}
+	if cfg.MaxRetries == 0 {
```

Comment:
I haven't looked through all the new code paths since I last touched this, but at the time the retries value was compared to number of attempts with a <=, so setting max retries to -1 (or any other negative number) would cause zero retries to happen.